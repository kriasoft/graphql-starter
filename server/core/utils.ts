/* SPDX-FileCopyrightText: 2014-present Kriasoft */
/* SPDX-License-Identifier: MIT */

import { fromGlobalId as parse } from "graphql-relay";
import { Knex } from "knex";
import { customAlphabet } from "nanoid";

// An alphabet for generating short IDs.
const alphabet = "0123456789abcdefghijklmnopqrstuvwxyz";

// Creates a function that generates a short ID of specified length.
export function createNewId(
  /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
  db: Knex<any, unknown[]>,
  table: string,
  size: number,
): (unique: boolean) => Promise<string> {
  const generateId = customAlphabet(alphabet, size);

  return async function newId(unique = true) {
    let id = generateId();

    // Ensures that the generated ID is unique
    while (unique) {
      const record = await db.table(table).where({ id }).first(db.raw(1));

      if (record) {
        console.warn(`Re-generating new ${table} ID.`);
        id = generateId();
      } else {
        break;
      }
    }

    return id;
  };
}

/**
 * Converts (Relay) global ID into a raw database ID.
 */
export function fromGlobalId(globalId: string, expectedType: string): string {
  const { id, type } = parse(globalId);

  if (expectedType && type !== expectedType) {
    throw new Error(
      `Expected an ID of type '${expectedType}' but got '${type}'.`,
    );
  }

  return id;
}
