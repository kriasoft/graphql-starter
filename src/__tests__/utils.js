/**
 * Node.js API Starter Kit (https://reactstarter.com/nodejs)
 *
 * Copyright Â© 2016-present Kriasoft, LLC. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

/* eslint-env jest */

import { mapTo, mapToMany, mapToValues } from '../utils';

describe('utils', () => {
  test('mapTo()', () => {
    const result = mapTo([1, 2], x => x.id, 'Test')([
      { id: 2, name: 'b' },
      { id: 1, name: 'a' },
    ]);
    expect(result).toMatchSnapshot();
  });

  test('mapToMany()', () => {
    const result = mapToMany([1, 2], x => x.id, 'Test')([
      { id: 2, name: 'b' },
      { id: 1, name: 'a' },
      { id: 1, name: 'c' },
    ]);
    expect(result).toMatchSnapshot();
  });

  test('mapToValues()', () => {
    const result = mapToValues([1, 2, 3], x => x.id, x => x.name)([
      { id: 2, name: 'b' },
      { id: 1, name: 'a' },
      { id: 3, name: 'c' },
    ]);
    expect(result).toMatchSnapshot();
  });
});
