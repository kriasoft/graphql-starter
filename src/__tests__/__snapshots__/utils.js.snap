// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils mapTo() 1`] = `
Array [
  Object {
    "__type": "Test",
    "id": 1,
    "name": "a",
  },
  Object {
    "__type": "Test",
    "id": 2,
    "name": "b",
  },
]
`;

exports[`utils mapToMany() 1`] = `
Array [
  Array [
    Object {
      "__type": "Test",
      "id": 1,
      "name": "a",
    },
    Object {
      "__type": "Test",
      "id": 1,
      "name": "c",
    },
  ],
  Array [
    Object {
      "__type": "Test",
      "id": 2,
      "name": "b",
    },
  ],
]
`;

exports[`utils mapToValues() 1`] = `
Array [
  "a",
  "b",
  "c",
]
`;
